File: src/main/java/net/exohayvan/dissolver_enhanced/common/analytics/PostHogCaptureClient.java
package net.exohayvan.dissolver_enhanced.common.analytics;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.BiConsumer;

public final class PostHogCaptureClient implements AutoCloseable {
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(5);

    private final URI endpoint;
    private final String apiKey;
    private final HttpClient httpClient;
    private final ExecutorService executor;
    private final BiConsumer<String, Throwable> warningLogger;

    public PostHogCaptureClient(String endpoint, String apiKey, BiConsumer<String, Throwable> warningLogger) {
        this.endpoint = URI.create(Objects.requireNonNull(endpoint, "endpoint"));
        this.apiKey = Objects.requireNonNull(apiKey, "apiKey");
        this.warningLogger = warningLogger;
        this.executor = Executors.newSingleThreadExecutor(task -> {
            Thread thread = new Thread(task, "dissolver-enhanced-posthog");
            thread.setDaemon(true);
            return thread;
        });
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(REQUEST_TIMEOUT)
            .executor(executor)
            .build();
    }

    public CompletableFuture<Void> capture(String event, String distinctId, Map<String, ?> properties) {
        Map<String, Object> eventProperties = new LinkedHashMap<>();
        if (properties != null) {
            eventProperties.putAll(properties);
        }
        eventProperties.put("distinct_id", distinctId);

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("api_key", apiKey);
        body.put("event", event);
        body.put("properties", eventProperties);
        body.put("timestamp", Instant.now().toString());

        HttpRequest request = HttpRequest.newBuilder(endpoint)
            .timeout(REQUEST_TIMEOUT)
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(PostHogJson.toJson(body)))
            .build();

        return httpClient.sendAsync(request, HttpResponse.BodyHandlers.discarding())
            .thenAccept(response -> {
                int statusCode = response.statusCode();
                if (statusCode < 200 || statusCode >= 300) {
                    logWarning("PostHog capture failed with HTTP " + statusCode + ".", null);
                }
            })
            .exceptionally(exception -> {
                logWarning("PostHog capture failed.", exception);
                return null;
            });
    }

    @Override
    public void close() {
        executor.shutdown();
    }

    private void logWarning(String message, Throwable throwable) {
        if (warningLogger != null) {
            warningLogger.accept(message, throwable);
        }
    }

}

File: stardewvalley/src/Analytics/PostHogCaptureClient.cs
using System.Text;
using System.Text.Json;

namespace DissolverEnhanced.StardewValley.Common.Analytics;

public sealed class PostHogCaptureClient : IDisposable
{
    private static readonly TimeSpan RequestTimeout = TimeSpan.FromSeconds(5);
    private readonly HttpClient httpClient;
    private readonly string endpoint;
    private readonly string apiKey;
    private readonly Action<string, Exception?> warningLogger;

    public PostHogCaptureClient(string endpoint, string apiKey, Action<string, Exception?> warningLogger)
    {
        this.endpoint = endpoint;
        this.apiKey = apiKey;
        this.warningLogger = warningLogger;
        httpClient = new HttpClient
        {
            Timeout = RequestTimeout
        };
    }

    public Task CaptureAsync(string eventName, string distinctId, IReadOnlyDictionary<string, object?>? properties)
    {
        Dictionary<string, object?> eventProperties = properties == null
            ? new Dictionary<string, object?>()
            : new Dictionary<string, object?>(properties);
        eventProperties["distinct_id"] = distinctId;

        Dictionary<string, object?> body = new()
        {
            ["api_key"] = apiKey,
            ["event"] = eventName,
            ["properties"] = eventProperties,
            ["timestamp"] = DateTimeOffset.UtcNow.ToString("O")
        };

        return SendAsync(endpoint, body);
    }

    public Task SetUserPropertiesAsync(string endpoint, string distinctId, IReadOnlyDictionary<string, object?> properties)
    {
        Dictionary<string, object?> body = new()
        {
            ["api_key"] = apiKey,
            ["event"] = "$set",
            ["distinct_id"] = distinctId,
            ["properties"] = new Dictionary<string, object?> { ["$set"] = properties },
            ["timestamp"] = DateTimeOffset.UtcNow.ToString("O")
        };

        return SendAsync(endpoint, body);
    }

    private async Task SendAsync(string requestEndpoint, IReadOnlyDictionary<string, object?> body)
    {
        try
        {
            string json = JsonSerializer.Serialize(body);
            using StringContent content = new(json, Encoding.UTF8, "application/json");
            using HttpResponseMessage response = await httpClient.PostAsync(requestEndpoint, content).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                warningLogger($"PostHog capture failed with HTTP {(int)response.StatusCode}.", null);
            }
        }
        catch (Exception exception)
        {
            warningLogger("PostHog capture failed.", exception);
        }
    }

    public void Dispose()
    {
        httpClient.Dispose();
    }
}
